cyipopt/scipy_interface.py

```diff
@@ -127,30 +159,49 @@ def gradient(self, x, **kwargs):
 
     def constraints(self, x):
         con_values = []
-        for fun, args in zip(self._constraint_funs, self._constraint_args):
-            con_values.append(fun(x, *args))
+        for i, (fun, args) in enumerate(
+                zip(self._constraint_funs, self._constraint_args)):
+            if self._constraint_funs_with_jacs[i]:
+                con_values.append(
+                    self.__evaluate_con_fun_with_jac(i, fun, x, *args)[0])
+            else:
+                con_values.append(fun(x, *args))
         return np.hstack(con_values)
 
     def jacobian(self, x):
         con_values = []
-        for fun, args in zip(self._constraint_jacs, self._constraint_args):
-            con_values.append(fun(x, *args))
+        for i, (fun, jac, args) in enumerate(
+                zip(self._constraint_funs, self._constraint_jacs, self._constraint_args)):
+            if self._constraint_funs_with_jacs[i]:
+                con_values.append(
+                    self.__evaluate_con_fun_with_jac(i, fun, x, *args)[1])
+            else:
+                con_values.append(jac(x, *args))
         return np.vstack(con_values)
 
-    def intermediate(
-            self,
-            alg_mod,
-            iter_count,
-            obj_value,
-            inf_pr,
-            inf_du,
-            mu,
-            d_norm,
-            regularization_size,
-            alpha_du,
-            alpha_pr,
-            ls_trials
-            ):
+    def __evaluate_con_fun_with_jac(self, i, con_fun, x, *args):
+        if self.last_x is None or not np.all(self.last_x == x):
+            self.last_x = x
+            self.last_con_values[i] = con_fun(x, *args)
+        return self.last_con_values[i]
```

Comment:
I personally don't think that it's clear what is being returning whenever this function is called. In the two cases it is called, the return value is index directly. Should this be split in to two separate methods/functions/closures that make the return value obvious?

Reply:
I agree it is a bit cumbersome. Does it help to add a few comments? The function returns a tuple `(fun_val, jac_val)` containing the function value `fun_val` and the evaluated jacobian `jac_val` of the `i`-th constraint.

Maybe something like

```  python
def _evaluate_con_fun_with_jac(self, i, con_fun, x, *args):
    if self.last_x is None or not np.all(self.last_x == x):
        self.last_x = x
        self.last_con_fun_values[i], self.last_con_jac_values[i] = con_fun(x, *args)
    return self.last_con_fun_values[i], self.last_con_jac_values[i]
```

makes it easier to understand at the first sight?

Reply:
Writing code that is clear in itself is significantly preferable to having to explain the code through comments. As this function is only being called in two places with one accessing the first return value and the other accessing the second, I still believe that the should be split in to two separate methods. One called `_evaluate_last_con_fun_value` and the other `_evaluate_last_con_jac_value` or similar.

Reply:
Since the same argument applies for `evaluate_fun_with_grad()`, shouldn't we then also split this method?

Reply:
Always amusing when PR code review flags up inconsistencies elsewhere in your codebase! Truthfully, yes, I think we should also split `evaluate_fun_with_grad()`. Or even better, refactor the relevant parts of it into `objective()` and `gradient()`. However, as `evaluate_fun_with_grad()` has up until this point been a public method, it will need to be deprecated correctly if being removed. In that case, probably best to ignore all of my comments relating to this and instead will open an issue regarding this so it can be fixed in a separate PR. Thanks @jhelgert! :)